OS I/O: Raw Devices

vyzo opened the pull request:
Cherry picked from #271.

Adds raw devices, currently implemented for POSIX only, so that special devices can be used to perform userland read/write operations while integrating with the Gambit scheduler.

vyzo commented:
@feeley any progress with the generic port hierarchy?
I can do them if you don't have the time!

vyzo commented:
I discovered a bug in the `___raw_device_select_virt` -- it fails to handle for_writing correctly; fixing.

feeley commented:
Halfway done...

feeley commented:
Waitable ports are now implemented by commit 05cec0edbd542fe4da76a78cced2969612199216.

vyzo commented:
Excellent! I will merge and update the raw-device-ports to be descendants of waitable-ports.

By the way, how do you generate the type-ids for the structs? Is it a random uuid or something structured?
I used a hash to generate it for raw-device-ports.

vyzo commented:
btw, `___PROCESS_DEVICE_KIND` disagrees with `(macro-process-kind)`.
I assume the two should be the same?
```
os_io.c:
#define ___PROCESS_DEVICE_KIND    (___PIPE_DEVICE_KIND+131072)

_io#.scm:
(##define-macro (macro-process-kind)     (+ 31 64))
```

vyzo commented:
unless the process is just a pipe in userland, in which case it's ok:
```
os_io.c:
#define ___PIPE_DEVICE_KIND       (___DEVICE_KIND+64)
```

vyzo commented:
Ok, merged! I'll do some casual testing and update #273 as well.

vyzo commented:
seems like my implementation of wait is not correct - `##wait-device` expects a device-port, so i'll have to call `##wait-for-io!` directly.

vyzo commented:
ok, verified that all port operations work as intended, this should be ready for merge!

vyzo commented:
Updates for review comments in #273.

The only sticking point left is the utility of `##open-raw-device` and its name.
For now it takes a device (not an fd :), opens a raw device and calls `##make-raw-device-port`.

We can drop the function altogether if you think the caller should be doing the actual opening of the device (ie if you think it's too POSIX specific)

vyzo commented:
Thinking more about the issue, is your concern how to obtain the actual device object that gets passed to `##open-raw-device` or the fact that the current implementation of `##os-device-open-raw` is only implemented for posix?
If the concern is the latter, then we should drop the function and keep just `##make-raw-device-port`. It's nicely convenient, but nothing that can't be done in a library.

vyzo commented:
On further thought, I will simply remove the function altogether.

vyzo commented:
@feeley any thoughts? I have addressed your review comments, both here and in #273.

I'd like to get this merged soon, as I already have libraries ready in a big Gerbil stdlib branch.
They provide the full POSIX socket API (nonblocking), epoll, and also high level network programming 
facilities that automate i/o multiplexing through epoll (and soon kqueue).

vyzo commented on lib/os_io.c:
```diff
@@ -8901,6 +8901,258 @@ ___SCMOBJ options;)
+}
+#endif
+
+___SCMOBJ ___os_device_raw_open
```
`open_predefined` tries to check the file descriptor type with fstat, which fails to take into account non-file or socket file descriptors. It also results in a hard crash with epoll file descriptors.

feeley replied:
I see, so it makes sense to have a function that does no device kind detection.  As far as I can see, this would mean that the methods should have the same implementation as for the "file device" kind.  So either, copy/paste or reuse the file device code (probably you need to copy/paste because the structure offsets are different).

vyzo replied:
Many methods are very similiar, in fact I based my implementation on the file device implementation.

feeley replied:
I keep thinking that it is wrong to expose this function in Scheme since is is POSIX specific. It is OK however to make it available to C code, so that a device can be constructed from a fd using the FFI. The "index" parameter should be "fd" because that's what it is.  The function should be called ```___os_device_raw_open_from_fd```.

vyzo replied:
I will rename.

The exposure is very minimal, there is just a very slim interface through `##os-device-open-raw`.
It's convenient to have it available to Scheme code, so that we can write simpler ffi code that constructs the fds and then calls the scheme function outside the C code. That way we can compose ffi constructor functions with a generic functions that does the `##open-raw-device`.

vyzo commented on lib/_kernel.scm:
```diff
@@ -4279,6 +4279,12 @@ end-of-code
             scheme-object   ;; device
    "___os_device_stream_open_process"))
 
+(define-prim ##os-device-open-raw
```
see comment above.

feeley replied:
See comment below.  This functionality should not be exposed to the Scheme level.

vyzo replied:
Do you prefer the ffi code to have to define the primitive on its own?
Ie instead of having the ffi interface in _kernel.scm to have it in user code?
I don't have any problem with removing the function from Scheme code in this case, but see comment above per convenience.

vyzo replied:
I will remove the function if you feel strongly about it, and move it to the library.

feeley replied:
I guess if it is purely a ```##``` function it can stay with a ```-from-fd``` suffix.

vyzo commented on lib/os_io.c:
```diff
@@ -8901,6 +8901,258 @@ ___SCMOBJ options;)
+  return ___FIX(___NO_ERR);
+}
+
+___HIDDEN ___SCMOBJ ___device_raw_select_virt
```
You can't read from many of these devices -- eg `epoll` and `kqueue` file descriptors.

feeley replied:
But since for some devices you can, it should be supported. If reading is not supported, then the "direction" should not contain the "in" direction, so all should be fine.

vyzo replied:
To clarify: the direction is still in, as they are waitable for input and have an open direction, but they can't be predefined ports because you can't do actual reads.

feeley replied:
Please make a copy of ```___device_file_select_raw_virt``` including the WIN32 part, because then it will probably work as-is on Windows.

vyzo replied:
ok, will do.

vyzo commented on lib/_io.scm:
```diff
@@ -8344,6 +8344,143 @@
+      ;; access to the port.
+
+      (##declare (not interrupts-enabled))
+      (##list uname (macro-raw-device-port-device port)))
```
I use it as a means for typing raw devices, and also having a nice display of the port.
I could call it name and make it non-optional, so that we are consistent and don't default in raw-device.
Finally, I can see the point about non-closures, I can simply add the name (better, type) to the `raw-device-port` type.

vyzo replied:
Added a type argument, and made the name return the type consed with the device for nice debugging output. 
No need to get the type from the name any more though, it's available directly from the port.

vyzo commented on lib/_io#.scm:
```diff
@@ -573,6 +574,31 @@
+
+  rdevice-condvar
+  wdevice-condvar
+  device
```
yes, will do.

vyzo replied:
called it type, as it is more representative of its usage

feeley replied:
What do you think of "id", which identifies the port?

vyzo replied:
Sure, even better; name also implies strongly that it's a symbol

vyzo commented:
Updated per review comments.

vyzo commented:
smal brainfart in usage of id field, fixing.

feeley commented on lib/os_io.c:
```diff
@@ -8901,6 +8901,263 @@ ___SCMOBJ options;)
+
+  fd = ___INT(index);
+
+  if ((e = ___device_raw_setup_from_fd
```
Indent properly... open paren for parameters should be 2 spaces indented from the ```___device_...```.

vyzo replied:
do you have an emacs mode for your indentation style?

feeley replied:
Nope.

feeley commented on lib/os_io.c:
```diff
@@ -8901,6 +8901,263 @@ ___SCMOBJ options;)
+      != ___FIX(___NO_ERR))
+    return e;
+
+  if ((e = ___NONNULLPOINTER_to_SCMOBJ
```
Same indent style should be applied here.

feeley commented:
Seems like ```___device_raw_select_virt``` should be called ```___device_raw_select_raw_virt``` to be consistent with the other device kinds.

vyzo commented:
ok, will rename.

vyzo commented:
ok, updated! Sorry for the indentation mismatch, i tend to just follow emacs in GNU style.

vyzo commented:
Seems like the git CR interface is hiding some of the commits in the update, but they are there in the commits page: https://github.com/gambit/gambit/pull/272/commits

vyzo commented:
actually there is an accident from the copy paste and it's a wrong cast in `___device_raw_select_raw_virt` -- fixing.

vyzo commented:
fixed in situ in the last commit and did a force push.

vyzo commented:
ok, now we should be good -- waiting for the build to finish to run some tests, but it looks good to me.

vyzo commented:
testing successful.

adam-7 commented:
@vyzo , what's a raw device, what are their usecases? Can you provide a short example script that illustrates their utility here?

vyzo commented:
Things like sockets, epoll, inotify and kquque file descriptors.
With raw devices we can open these OS-specific devices and integrate them with the Gambit scheduler.

adam-7 commented:
Ah!

1)
By "raw device", do you just mean that you have created a facility whereby the user can allocate his own OS file descriptor of any type that the OS can produce,

(I have not studied epoll/inotify/kqueue so it's not in my awareness that they produce their own FD:s, but I understand that's what you're saying, so they fit the picture too;)

and, what you have done here is to provide a facility where the user can create a Gambit port object that functions only to represent such a file descriptor, so that the user may do |##device-port-wait-for-input!| and |##device-port-wait-for-output!| on that port object?

2)
I guess such a raw port object is actually totally passive, as in, Gambit will not do any operations on it such as OS write, read, close operations, but they function only to make Gambit include that FD in Gambit's select/poll() call, during the ##device-port-wait-for-input/output! call?

3)
Are any other operations than ##device-port-wait-for-input/output! supported on those raw ports?

4)
Are the raw ports documented in the manual now or do you keep these out of the manual due to being too low level?

5)
Probably the raw ports will be used for pretty low level stuff, right?

6)
Except for this and making Gambit's epoll() support unlimited number of ports (whereas previously there was an ~8000 FD:s limit and exceeding that led to SIGSEGV), have you done any other major addition now?

Looks great! :) Thanks

feeley commented:
This is just a heads up that the API for ```##make-raw-device-port``` will be changing to accommodate other types of raw-devices.  In particular what used to be called the "id" is now called the "type", and the "name" of the port is an explicit parameter (so it is easy to choose a custom name). The last field is a "specific" field that can be used to store anything.

Vyzo, here is how the old API can be emulated:

```
(define (##vyzos-make-raw-device-port raw-device device id direction)
  (##make-raw-device-port direction raw-device id (##list id device) device))
```

To access the "device" from the port, use ```(macro-raw-device-port-specific port)```.

adam-7 commented:
Like what types?

vyzo commented:
thanks marc!

feeley commented:
I'm implementing UDP ports and it is convenient to build them on top of raw-device ports.  By the way, the API has changed again to: ```(##make-raw-device-port direction device kind id specific)```.  The third parameter is a numeric "kind", for example ```(macro-raw-device-kind)```.  This simplifies type testing.

vyzo commented:
Can the kind be an arbitrary object (eg a symbol) and not just a numeric kind in the new API?
That would be much better for identifying user ports.
